### PR TITLE
Drastically Improve Performance

### PR DESCRIPTION
**Pull request by issafram**
Use multiple threads to process faster.

### CODE REVIEW

**Comment by bradwestness** [comment]
Do you have any benchmarks to actually show the performance of this versus the single-threaded version? I'm a little concerned that all the .ToList() calls are going to cancel out any parrallelization wins by enumerating all the collections a bunch of extra times.

**Comment by issafram** [comment]
The `.ToList()` calls were actually so the library wouldn't try to enumerate multiple times when it isn't a simple collection.  Like for example, if this was an `IQueryable` that was being passed in from a database query.  The suggestion was made from ReSharper and I thought it was a good one.

I can add some performance benchmarks to show the differences with the multi-threaded changes.

**Comment by bradwestness** [comment]
Yeah, I'd be interested to see how much of a boost it gets with a big data set. Would be cool to see if it's a big improvement! :smile:

**Comment by issafram** [comment]
Some benchmarks using code from https://gist.github.com/issafram/b4bc98de575235456786

With the **current processing logic in production (1.3.4.1)**, we see results like these:
Time for 1000 rows on execution 1: 00:00:00.1144958
Time for 1000 rows on execution 2: 00:00:00.0402236
Time for 1000 rows on execution 3: 00:00:00.0342082
Time for 1000 rows on execution 4: 00:00:00.0391841
Time for 1000 rows on execution 5: 00:00:00.0421885
Average time for 1000 rows with 5 executions: 00:00:00.0540600

Time for 10000 rows on execution 1: 00:00:00.6753164
Time for 10000 rows on execution 2: 00:00:00.6652991
Time for 10000 rows on execution 3: 00:00:00.6625778
Time for 10000 rows on execution 4: 00:00:00.6550143
Time for 10000 rows on execution 5: 00:00:00.6464313
Average time for 10000 rows with 5 executions: 00:00:00.6609278
	
Time for 100000 rows on execution 1: 00:00:32.8842859
Time for 100000 rows on execution 2: 00:00:32.5784661
Time for 100000 rows on execution 3: 00:00:32.5285126
Time for 100000 rows on execution 4: 00:00:32.5694837
Time for 100000 rows on execution 5: 00:00:32.4996665
Average time for 100000 rows with 5 executions: 00:00:32.6120830


With the **new processing logic contained in this PR**, we see results like the following:
Time for 1000 rows on execution 1: 00:00:00.0905472
Time for 1000 rows on execution 2: 00:00:00.0237432
Time for 1000 rows on execution 3: 00:00:00.0170057
Time for 1000 rows on execution 4: 00:00:00.0219092
Time for 1000 rows on execution 5: 00:00:00.0220426
Average time for 1000 rows with 5 executions: 00:00:00.0350496

Time for 10000 rows on execution 1: 00:00:00.2325847
Time for 10000 rows on execution 2: 00:00:00.2193486
Time for 10000 rows on execution 3: 00:00:00.2311276
Time for 10000 rows on execution 4: 00:00:00.2297525
Time for 10000 rows on execution 5: 00:00:00.2263789
Average time for 10000 rows with 5 executions: 00:00:00.2278385

Time for 100000 rows on execution 1: 00:00:02.2678210
Time for 100000 rows on execution 2: 00:00:02.3526032
Time for 100000 rows on execution 3: 00:00:02.2568745
Time for 100000 rows on execution 4: 00:00:02.2974694
Time for 100000 rows on execution 5: 00:00:02.2818754
Average time for 100000 rows with 5 executions: 00:00:02.2913287

Time for 500000 rows on execution 1: 00:00:11.6423330
Time for 500000 rows on execution 2: 00:00:11.6322395
Time for 500000 rows on execution 3: 00:00:11.7896660
Time for 500000 rows on execution 4: 00:00:12.7939964
Time for 500000 rows on execution 5: 00:00:12.2637361
Average time for 500000 rows with 5 executions: 00:00:12.0243942

Time for 1000000 rows on execution 1: 00:00:25.1888954
Time for 1000000 rows on execution 2: 00:00:25.7263118
Time for 1000000 rows on execution 3: 00:00:25.7460554
Time for 1000000 rows on execution 4: 00:00:27.1094301
Time for 1000000 rows on execution 5: 00:00:25.3239448
Average time for 1000000 rows with 5 executions: 00:00:25.8189275



So basically,  this change can process 1 *million* rows **faster** than the current logic can process even one hundred *thousand* rows.
